In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [44]:
ds = pd.Series([1.0, 3.5, 4.6, 7.8])

In [3]:
ds

0    1.0
1    3.5
2    4.6
3    7.8
dtype: float64

In [4]:
ds[0]

1.0

In [45]:
import numpy as np

In [6]:
np.random.randn(10)

array([ 0.6193976 , -0.43738291, -0.34687977,  0.23278036,  1.1854454 ,
       -0.57256875,  1.75208346,  0.03556778, -0.16074252, -0.09515212])

In [7]:
np.random.randn(10)*5 + 25

array([32.85284154, 34.69530231, 24.97448679, 25.42965367, 15.79312401,
       38.68741594, 26.41131575, 25.77922713, 28.37656181, 21.26185979])

In [4]:
df = pd.DataFrame( {"names":["Samudra", "Sagar", "Alice", "Bob", "Ekman", "Munk", "Sindhu", "Tharp", "McBoaty", "Earle"], 
                    "age":np.random.randn(10)*10 + 30, 
                    "weight":np.random.randn(10)*25 + 75, 
                    "university":["Unseen", "Mangalore", "Goa", "Mangalore", "Unseen", "Unseen", "Goa", "Mangalore", "Unseen", "Unseen"]}, 
                   index=["A", "B", "C", "D", "E", "F", "G", "H", "I", "J"])

In [5]:
df

names        age      weight university
A  Samudra  38.795587   78.024792     Unseen
B    Sagar  25.828514   81.853686  Mangalore
C    Alice  35.622627   86.470429        Goa
D      Bob  24.746744   51.886843  Mangalore
E    Ekman  29.460587   80.617955     Unseen
F     Munk  32.000563  105.504541     Unseen
G   Sindhu  35.086820   70.798316        Goa
H    Tharp  40.644060   48.891555  Mangalore
I  McBoaty  30.503795   81.304399     Unseen
J    Earle  42.225321   46.230288     Unseen

In [48]:
df.columns

Index(['names', 'age', 'weight', 'university'], dtype='object')

In [49]:
df.index

Index(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'], dtype='object')

In [50]:
df.loc[["A","J"]]

names        age     weight university
A    Jim  39.413893  63.998811     Unseen
J  Earle  30.517096  23.216171     Unseen

In [51]:
df.loc[["A","J"], "age"]

A    39.413893
J    30.517096
Name: age, dtype: float64

In [52]:
df.iloc[[0,9]]

names        age     weight university
A    Jim  39.413893  63.998811     Unseen
J  Earle  30.517096  23.216171     Unseen

#### Defining your own index, rather than the default index of 0...N may be useful when you need to organize data logically, for example time series data can use the time as index
But luckily, you also have access to the 0...N index by using the "iloc" selector

In [53]:
df["age"]

A    39.413893
B    32.020081
C    41.591701
D    44.017436
E    25.221935
F    20.949035
G    23.219359
H    24.887251
I    31.030534
J    30.517096
Name: age, dtype: float64

In [54]:
df["weight"].values # .values returns the column as an array

array([ 63.99881067,  68.92419793, 101.95900834,  78.65857272,
        73.73933644,  79.80366156,  84.20366488, 114.82115745,
       120.29430863,  23.21617103])

In [55]:
df.describe()

age      weight
count  10.000000   10.000000
mean   31.286832   80.961889
std     8.063604   27.830657
min    20.949035   23.216171
25%    24.970922   70.127983
50%    30.773815   79.231117
75%    37.565440   97.520172
max    44.017436  120.294309

In [56]:
df["university"].unique()

array(['Unseen', 'Gothenburg', 'Lund'], dtype=object)

In [6]:
sel_UU = (df["university"] == "Unseen")
df.loc[sel_UU] # You can use boolean selectors to select slices of the dataframe

names        age      weight university
A  Samudra  38.795587   78.024792     Unseen
E    Ekman  29.460587   80.617955     Unseen
F     Munk  32.000563  105.504541     Unseen
I  McBoaty  30.503795   81.304399     Unseen
J    Earle  42.225321   46.230288     Unseen

In [9]:
df.groupby("university").count()

names  age  weight
university                    
Goa             2    2       2
Mangalore       3    3       3
Unseen          5    5       5

In [10]:
df.columns

Index(['names', 'age', 'weight', 'university'], dtype='object')

In [13]:
df.loc[:, ["age", "weight", "university"]].groupby("university").mean()

age     weight
university                      
Goa         35.354724  78.634372
Mangalore   30.406439  60.877361
Unseen      34.597171  78.336395

In [15]:
df.loc[:, ["age", "weight", "university"]].groupby("university").std()

age     weight
university                     
Goa         0.378873  11.081857
Mangalore   8.882523  18.227660
Unseen      5.605755  21.112922

In [16]:
df.loc[df["age"] > 25]

names        age      weight university
A  Samudra  38.795587   78.024792     Unseen
B    Sagar  25.828514   81.853686  Mangalore
C    Alice  35.622627   86.470429        Goa
E    Ekman  29.460587   80.617955     Unseen
F     Munk  32.000563  105.504541     Unseen
G   Sindhu  35.086820   70.798316        Goa
H    Tharp  40.644060   48.891555  Mangalore
I  McBoaty  30.503795   81.304399     Unseen
J    Earle  42.225321   46.230288     Unseen

In [17]:
df.loc[df["age"] > 25, "names"]

A    Samudra
B      Sagar
C      Alice
E      Ekman
F       Munk
G     Sindhu
H      Tharp
I    McBoaty
J      Earle
Name: names, dtype: object

In [18]:
df.loc[df["age"] > 25, "weight"] * 1.25

A     97.530991
B    102.317108
C    108.088036
E    100.772444
F    131.880676
G     88.497895
H     61.114443
I    101.630498
J     57.787860
Name: weight, dtype: float64

In [19]:
df.loc[df["age"] > 25, "weight"] = df.loc[df["age"] > 25, "weight"] * 1.25

In [20]:
df

names        age      weight university
A  Samudra  38.795587   97.530991     Unseen
B    Sagar  25.828514  102.317108  Mangalore
C    Alice  35.622627  108.088036        Goa
D      Bob  24.746744   51.886843  Mangalore
E    Ekman  29.460587  100.772444     Unseen
F     Munk  32.000563  131.880676     Unseen
G   Sindhu  35.086820   88.497895        Goa
H    Tharp  40.644060   61.114443  Mangalore
I  McBoaty  30.503795  101.630498     Unseen
J    Earle  42.225321   57.787860     Unseen

#### You can define your own functions and perform operations on grouped dataframes using "apply"

Let's say that a magical accident ages everyone at Unseen University by 25 years, we can update our dataframe by grouping it by "university" and then adding 25 years to the ages of those in UU.

In [66]:
def increase_age(gdf, addYears=25):
    if gdf.university[0] == "Unseen": # conditional statement to perform the operation only on Unseen University members
        gdf.loc[:, "age"] = gdf["age"] + addYears
    return gdf

In [67]:
df = df.groupby("university").apply(increase_age)

In [68]:
df

names        age      weight  university
A      Jim  64.413893   79.998513      Unseen
B     Jane  32.020081   86.155247  Gothenburg
C    Alice  41.591701  127.448760        Lund
D      Bob  44.017436   98.323216  Gothenburg
E    Ekman  50.221935   92.174171      Unseen
F     Munk  45.949035   79.803662      Unseen
G   Nansen  23.219359   84.203665        Lund
H    Tharp  24.887251  114.821157  Gothenburg
I  McBoaty  56.030534  150.367886      Unseen
J    Earle  55.517096   29.020214      Unseen

In [69]:
df.groupby(pd.cut(df["age"], [20,30,40,50,60,70])).count()

names  age  weight  university
age                                     
(20, 30]      2    2       2           2
(30, 40]      1    1       1           1
(40, 50]      3    3       3           3
(50, 60]      3    3       3           3
(60, 70]      1    1       1           1

pd.cut() function allows us to bin the data. You pass the column followed by the bins and then call a function that operates on the grouped dataframe.

### Exercise 05: Practise groupby operations

1. Groupby university name and use the following functions: count, mean, std, describe
1. Groupby weight, binned in (0,50], (50, 100], (100,150], (150, 200] bins and call the count and mean functions.
1. Can you draw a flowchart describing the groupby() and apply() operation? Discuss this with your group members!